### ISSUE TITLE
Fix logit loss

### ISSUE THREAD

**Issue by mnwright** [PlantedML/randomPlantedForest#40]
https://github.com/PlantedML/randomPlantedForest/blob/3d3b10b222a2962f2d64dd7e0dc394a8974fb4c9/src/randomPlantedForest.cpp#L2053-L2058
If `W` and `W_new` are the same, why then treat them separately below?

https://github.com/PlantedML/randomPlantedForest/blob/3d3b10b222a2962f2d64dd7e0dc394a8974fb4c9/src/randomPlantedForest.cpp#L2068-L2069
`W` and `W_new` are the same, so nothing happens?
And won't this give `log(0)` in some cases?

**Comment by mnwright** [comment]
OK I see, `W` and `W_new` are copies of `*split.W`. Can we do that more efficiently?

**Comment by jemus42** [comment]
Regarding logit loss (multiclass(!)), I am reasonably certain that something is broken, at least it gives implausibly bad results compared to exponential and L1:

``` r
xdf <- na.omit(palmerpenguins::penguins)

rpfitlogit <- randomPlantedForest::rpf(species ~ ., data = xdf, loss = "logit", ntrees = 50, max_interaction = 3)
predict(rpfitlogit, xdf, type = "prob") |>
  cbind(xdf[, "species"]) |> 
  head(20)
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo species
#> 1     0.3333333       0.3333333    0.3333333  Adelie
#> 2     0.3333333       0.3333333    0.3333333  Adelie
#> 3     0.3333333       0.3333333    0.3333333  Adelie
#> 4     0.3333333       0.3333333    0.3333333  Adelie
#> 5     0.3333333       0.3333333    0.3333333  Adelie
#> 6     0.3333333       0.3333333    0.3333333  Adelie
#> 7     0.3333333       0.3333333    0.3333333  Adelie
#> 8     0.3333333       0.3333333    0.3333333  Adelie
#> 9     0.3333333       0.3333333    0.3333333  Adelie
#> 10    0.3333333       0.3333333    0.3333333  Adelie
#> 11    0.3333333       0.3333333    0.3333333  Adelie
#> 12    0.3333333       0.3333333    0.3333333  Adelie
#> 13    0.3333333       0.3333333    0.3333333  Adelie
#> 14    0.3333333       0.3333333    0.3333333  Adelie
#> 15    0.3333333       0.3333333    0.3333333  Adelie
#> 16    0.3333333       0.3333333    0.3333333  Adelie
#> 17    0.3333333       0.3333333    0.3333333  Adelie
#> 18    0.3333333       0.3333333    0.3333333  Adelie
#> 19    0.3333333       0.3333333    0.3333333  Adelie
#> 20    0.3333333       0.3333333    0.3333333  Adelie
```

## Without logistic transformation = raw predictions

```r
predict(rpfitlogit, xdf, type = "numeric") |>
        cbind(xdf[, "species"]) |> 
        head(20)
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo species
#> 1             0               0            0  Adelie
#> 2             0               0            0  Adelie
#> 3             0               0            0  Adelie
#> 4             0               0            0  Adelie
#> 5             0               0            0  Adelie
#> 6             0               0            0  Adelie
#> 7             0               0            0  Adelie
#> 8             0               0            0  Adelie
#> 9             0               0            0  Adelie
#> 10            0               0            0  Adelie
#> 11            0               0            0  Adelie
#> 12            0               0            0  Adelie
#> 13            0               0            0  Adelie
#> 14            0               0            0  Adelie
#> 15            0               0            0  Adelie
#> 16            0               0            0  Adelie
#> 17            0               0            0  Adelie
#> 18            0               0            0  Adelie
#> 19            0               0            0  Adelie
#> 20            0               0            0  Adelie
```

# Exponential

```r
rpfitexp <- randomPlantedForest::rpf(species ~ ., data = xdf, loss = "exponential", ntrees = 50, max_interaction = 3)
predict(rpfitexp, xdf, type = "prob") |>
  cbind(xdf[, "species"]) |> 
  head(20)
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo species
#> 1     1.0000000    6.082873e-12 7.106864e-14  Adelie
#> 2     1.0000000    4.661963e-10 5.068072e-12  Adelie
#> 3     1.0000000    7.590473e-10 4.827989e-12  Adelie
#> 4     1.0000000    5.579261e-11 5.763339e-13  Adelie
#> 5     1.0000000    8.626273e-12 9.987598e-14  Adelie
#> 6     1.0000000    1.132616e-10 1.141771e-12  Adelie
#> 7     1.0000000    3.388306e-11 2.270850e-12  Adelie
#> 8     0.9999982    1.801812e-06 4.336772e-09  Adelie
#> 9     1.0000000    5.977995e-12 8.090228e-14  Adelie
#> 10    1.0000000    3.778767e-11 1.010726e-12  Adelie
#> 11    1.0000000    9.578875e-11 9.941101e-13  Adelie
#> 12    1.0000000    5.723994e-11 6.154511e-13  Adelie
#> 13    0.9970270    2.958427e-03 1.460202e-05  Adelie
#> 14    1.0000000    2.339126e-10 2.148345e-12  Adelie
#> 15    0.5857859    4.136132e-01 6.009288e-04  Adelie
#> 16    1.0000000    1.932036e-10 1.343189e-12  Adelie
#> 17    1.0000000    1.145846e-11 9.924348e-14  Adelie
#> 18    1.0000000    7.778022e-11 8.125688e-13  Adelie
#> 19    1.0000000    1.387532e-11 1.393543e-13  Adelie
#> 20    1.0000000    1.111520e-10 1.238551e-12  Adelie
```

## Without logistic transformation = raw predictions

```r
predict(rpfitexp, xdf, type = "numeric") |>
  cbind(xdf[, "species"]) |> 
  head(20)
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo species
#> 1   23.00124736      -2.8242966    -7.273883  Adelie
#> 2   18.58962309      -2.8967913    -7.418438  Adelie
#> 3   18.39393981      -2.6050172    -7.662651  Adelie
#> 4   21.05259738      -2.5567824    -7.129492  Adelie
#> 5   22.72962552      -2.7465831    -7.205222  Adelie
#> 6   20.18552133      -2.7157999    -7.312919  Adelie
#> 7   19.81425752      -4.2938485    -6.996609  Adelie
#> 8   10.90784306      -2.3188731    -8.348291  Adelie
#> 9   22.96236676      -2.8805691    -7.183168  Adelie
#> 10  20.60119577      -3.3978426    -7.019157  Adelie
#> 11  20.37910472      -2.6897712    -7.257824  Adelie
#> 12  20.96752947      -2.6162398    -7.148891  Adelie
#> 13   3.60629156      -2.2138285    -7.525082  Adelie
#> 14  19.74682112      -2.4292526    -7.119502  Adelie
#> 15  -0.08395016      -0.4319734    -6.966183  Adelie
#> 16  19.38371681      -2.9835596    -7.952258  Adelie
#> 17  21.97460291      -3.2176896    -7.966597  Adelie
#> 18  20.05065356      -3.2264804    -7.787922  Adelie
#> 19  21.48711622      -3.5137930    -8.114641  Adelie
#> 20  19.39195945      -3.5281631    -8.025119  Adelie
```

# L1

```r
rpfitl1 <- randomPlantedForest::rpf(species ~ ., data = xdf, loss = "L1", ntrees = 50, max_interaction = 3)
predict(rpfitl1, xdf, type = "prob") |>
  cbind(xdf[, "species"]) |> 
  head(20)
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo species
#> 1     1.0000000    0.000000e+00 0.000000e+00  Adelie
#> 2     0.9999094    9.051544e-05 9.197701e-08  Adelie
#> 3     0.9958636    3.956696e-03 1.797305e-04  Adelie
#> 4     0.9999101    8.976917e-05 9.197701e-08  Adelie
#> 5     0.9999999    0.000000e+00 9.197700e-08  Adelie
#> 6     1.0000000    0.000000e+00 0.000000e+00  Adelie
#> 7     0.9858727    0.000000e+00 1.412726e-02  Adelie
#> 8     0.9655242    3.430549e-02 1.702658e-04  Adelie
#> 9     0.9999999    0.000000e+00 9.197700e-08  Adelie
#> 10    0.9854313    0.000000e+00 1.456870e-02  Adelie
#> 11    0.9999101    8.976917e-05 9.197701e-08  Adelie
#> 12    0.9999101    8.976917e-05 9.197701e-08  Adelie
#> 13    0.8618094    1.065320e-01 3.165860e-02  Adelie
#> 14    1.0000000    0.000000e+00 0.000000e+00  Adelie
#> 15    0.9501284    1.284619e-02 3.702538e-02  Adelie
#> 16    0.9998595    1.394223e-04 1.088739e-06  Adelie
#> 17    0.9999989    0.000000e+00 1.088738e-06  Adelie
#> 18    0.9997922    2.077787e-04 0.000000e+00  Adelie
#> 19    0.9999971    0.000000e+00 2.890525e-06  Adelie
#> 20    0.9999989    0.000000e+00 1.088738e-06  Adelie
```

## Without normalizing transformation = raw predictions

```r
predict(rpfitl1, xdf, type = "numeric") |>
  cbind(xdf[, "species"]) |> 
  head(20)
#>    .pred_Adelie .pred_Chinstrap  .pred_Gentoo species
#> 1     1.0009483   -9.466285e-04 -1.709817e-06  Adelie
#> 2     0.9999094    9.051544e-05  9.197701e-08  Adelie
#> 3     0.9958636    3.956696e-03  1.797305e-04  Adelie
#> 4     0.9999101    8.976917e-05  9.197701e-08  Adelie
#> 5     1.0006846   -6.846784e-04  9.197701e-08  Adelie
#> 6     1.0001724   -1.706885e-04 -1.709817e-06  Adelie
#> 7     0.9880243   -2.182434e-03  1.415809e-02  Adelie
#> 8     0.9655242    3.430549e-02  1.702658e-04  Adelie
#> 9     1.0006861   -6.861709e-04  9.197701e-08  Adelie
#> 10    0.9964013   -1.113217e-02  1.473088e-02  Adelie
#> 11    0.9999101    8.976917e-05  9.197701e-08  Adelie
#> 12    0.9999101    8.976917e-05  9.197701e-08  Adelie
#> 13    0.8618094    1.065320e-01  3.165860e-02  Adelie
#> 14    1.0002436   -2.376674e-04 -5.887823e-06  Adelie
#> 15    0.9501284    1.284619e-02  3.702538e-02  Adelie
#> 16    0.9998595    1.394223e-04  1.088739e-06  Adelie
#> 17    1.0006354   -6.365178e-04  1.088739e-06  Adelie
#> 18    0.9997983    2.077799e-04 -6.096483e-06  Adelie
#> 19    1.0102117   -1.021454e-02  2.890533e-06  Adelie
#> 20    1.0006354   -6.365178e-04  1.088739e-06  Adelie
```

<sup>Created on 2022-11-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
